### PR TITLE
Partial Fix for Issue 16073 (part 3)

### PR DESCRIPTION
**Pull request by JackStouffer**

### CODE REVIEW

**Comment by JackStouffer** [comment]
Pinging as this is a part of a series of important fixes

@schveiguy @qznc @wilzbach

File: std/range/package.d [hunk 1]
```diff
@@ -8030,7 +8036,7 @@ if (isInputRange!Range)
     // of the range and then 1 for the rest, returns the index at
     // which the first 1 appears. Used internally by the search routines.
     private size_t getTransitionIndex(SearchPolicy sp, alias test, V)(V v)
-    if (sp == SearchPolicy.binarySearch && isRandomAccessRange!Range)
+    if (sp == SearchPolicy.binarySearch && isRandomAccessRange!Range && hasLength!Range)
```
Comment (wilzbach) [inline std/range/package.d]: afaik RandomAccess requires `hasLength`

Comment (schveiguy) [inline std/range/package.d]: Infinite ranges can be random access

**Comment by wilzbach** [comment]
@JackStouffer have a look at this line: https://github.com/dlang/phobos/blob/master/std/range/primitives.d#L906

```d
 enum bool isRandomAccessRange = 
...
        static assert(hasLength!R || isInfinite!R);
...
``

File: std/range/package.d (hunk 1 above)
Comment (wilzbach) [inline std/range/package.d]: > Infinite ranges can be random access

Oh just saw that - sorry.

**Comment by JackStouffer** [comment]
Ping @schveiguy

File: std/range/package.d [hunk 2]
```diff
@@ -6530,10 +6533,13 @@ struct Indexed(Source, Indices)
...
 
-        /// Ditto
-        typeof(this) opSlice(size_t a, size_t b)
+        static if (hasSlicing!Indices)
         {
-            return typeof(this)(_source, _indices[a..b]);
+            /// Ditto
+            typeof(this) opSlice(size_t a, size_t b)
+            {
+                return typeof(this)(_source, _indices[a..b]);
```
Comment (schveiguy) [inline std/range/package.d]: `a .. b`

File: std/range/package.d [hunk 3]
```diff
@@ -10021,7 +10027,7 @@ auto padRight(R, E)(R r, E e, size_t n) if (
                     b <= length,
                     "Attempting to slice using an out of bounds index on a padRight"
                 );
-                return Result((b <= data.length) ? data[a .. b] : data[a .. $],
+                return Result((b <= data.length) ? data[a .. b] : data[a .. data.length],
```
Comment (schveiguy) [inline std/range/package.d]: Need `hasLength!R` for this call as well.

**Comment by schveiguy** [comment]
LGTM after couple nits. It pains me to let all those `typeof(this)` get through, but probably a different PR for that :)

**Comment by JackStouffer** [comment]
@schveiguy Done

File: std/range/package.d [hunk 4]
```diff
@@ -3370,21 +3370,24 @@ struct Cycle(R)
...
-        /// ditto
-        auto opSlice(size_t i, size_t j)
-        in
-        {
-            assert(i <= j);
-        }
-        body
+        static if (hasSlicing!R)
         {
-            return this[i .. $].takeExactly(j - i);
```
Comment (9il) [inline std/range/package.d]: `takeExactly` may work with infinity ranges

Comment (JackStouffer) [inline std/range/package.d]: opDollar doesn't

Comment (9il) [inline std/range/package.d]: Length is not opDollar

Comment (andralex) [inline std/range/package.d]: @9il is right!

**Comment by andralex** [comment]
@JackStouffer push this over the edge please? Thx!

File: std/range/package.d (hunk 4 above)
Comment (JackStouffer) [inline std/range/package.d]: @9il @andralex Doesn't matter, infinite ranges are already excluded from this path on line 3289

**Comment by JackStouffer** [comment]
Ping @andralex @9il

File: std/range/package.d (hunk 3 above)
Comment (andralex) [inline std/range/package.d]: @JackStouffer ?

**Comment by andralex** [comment]
LGTM after you add hasLength

**Comment by JackStouffer** [comment]
@andralex `hasLength` is already there, see https://github.com/dlang/phobos/pull/4463/files#diff-43f17b39f02da9efd68008101ca4cd81R3289

**Comment by schveiguy** [comment]
Auto-merge toggled on

**Comment by JackStouffer** [comment]
Thanks!

**Pull request merged**